File: src/main/java/com/xpdustry/nucleus/version/MindustryVersion.java
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.version;

import java.util.Locale;
import java.util.Objects;

/// A Mindustry version.
public record MindustryVersion(int major, int build, int patch, Type type) {

    public MindustryVersion {
        if (major < 0) {
            throw new IllegalArgumentException("Major version must be positive");
        }
        if (build < 0) {
            throw new IllegalArgumentException("Build version must be positive");
        }
        if (patch < 0) {
            throw new IllegalArgumentException("Patch version must be positive");
        }
        Objects.requireNonNull(type, "type");
    }

    @Override
    public String toString() {
        final var name = this.type.name().toLowerCase(Locale.ROOT).replace('_', '-');
        return name + " v" + this.major + " " + this.build + (this.patch == 0 ? "" : "." + this.patch);
    }

    /// A Mindustry release channel.
    public enum Type {
        OFFICIAL,
        ALPHA,
        BLEEDING_EDGE,
        CUSTOM,
    }
}

File: src/main/kotlin/com/xpdustry/nucleus/version/MindustryVersion.kt
// SPDX-License-Identifier: GPL-3.0-only
package com.xpdustry.nucleus.version

/** A Mindustry version. */
data class MindustryVersion(val major: Int, val build: Int, val patch: Int, val type: Type) {
    override fun toString(): String {
        val name = this.type.name.lowercase().replace('_', '-')
        return name + " v" + this.major + " " + this.build + (if (this.patch == 0) "" else "." + this.patch)
    }

    /** A Mindustry release channel. */
    enum class Type {
        OFFICIAL,
        ALPHA,
        BLEEDING_EDGE,
        CUSTOM,
    }

    init {
        require(major >= 0) { "Major version must be positive" }
        require(build >= 0) { "Build version must be positive" }
        require(patch >= 0) { "Patch version must be positive" }
    }
}
